### PR TITLE
[VIS] Implement Matopi Golem

### LINKED ISSUE #10016: [Bug] Soldevi Sentry doesn't allow opponent to draw a card
[[Soldevi Sentry]]

Looks like the text was previously hacked on without confirming correct functionality. Probably could be refactored to a simple implementation.

### PR DESCRIPTION
**Pull request by Susucre**
Also, rework [[Soldevi Sentry]] and [[Skeleton Scavengers]].

With the reflexive trigger, there should be more feedback to identify if Soldevi Sentry still is bugged (but from my testing it is working properly here)
image

closes #10016

### CODE REVIEW

**Comment by Susucre** [comment]
[[Soldevi Sentry]] [[Skeleton Scavengers]] [[Matopi Golem]]

File: Mage.Sets/src/mage/cards/m/MatopiGolem.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        Permanent permanent = game.getPermanent(source.getSourceId());
+        if (permanent != null && permanent.regenerate(source, game)) {
+            game.fireReflexiveTriggeredAbility(
+                    new ReflexiveTriggeredAbility(
+                            new AddCountersSourceEffect(CounterType.M1M1.createInstance()),
+                            false
+                    ).setTriggerPhrase("When it regenerates this way, "),
+                    source
+            );
+            this.used = true;
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/m/MatopiGolem.java:73]: Why no call to `discard()` like in `RegenerateSourceEffect`?

Actually, why not call `super.apply()` and fire the reflexive trigger if that returns true?

File: Mage.Sets/src/mage/cards/m/MatopiGolem.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Override
+    public MatopiGolem copy() {
+        return new MatopiGolem(this);
+    }
+}
+
+class MatopiGolemEffect extends RegenerateSourceEffect {
+
+    public MatopiGolemEffect() {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/m/MatopiGolem.java:48]: access modifiers

File: Mage.Sets/src/mage/cards/s/SoldeviSentry.java [hunk 3]
```diff
@@ -45,37 +48,69 @@ public SoldeviSentry copy() {
...
+            );
             this.used = true;
             return true;
         }
         return false;
     }
-
 }
+
+class SoldeviSentryEffect extends OneShotEffect {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/s/SoldeviSentry.java:87]: is there a reason not to use `new DrawCardTargetEffect(1, true)` ?

**Review by xenohedron — COMMENTED** [review]
So the three separate effects are okay because this is only used on three old cards and unlikely to be used in future designs. Still, it might be more elegant to use common logic with one of the following options:
1. new class `RegenerateSourceWithReflexiveEffect extends RegenerateSourceEffect` that takes a `ReflexiveTriggeredAbility`
2. helper method `withReflexiveTrigger()` in `RegenerateSourceEffect` for the same purpose

though I guess maybe you need another param for the target pointer, not sure

It looks like the logic for choosing which replacement effect to resolve first depends on the text, so should be fine even with a common class.

A unit test would be nice, also not required.

File: Mage.Sets/src/mage/cards/s/SoldeviSentry.java (hunk 3 above)
Comment (Susucre) [reply Mage.Sets/src/mage/cards/s/SoldeviSentry.java:87]: Thanks, i'll use that.
I too quickly searched for "target player may draw", did not find any match, and assumed the common effect would not allow for optional draw of non-source-controller player.

File: Mage/src/main/java/mage/abilities/effects/common/RegenerateSourceWithReflectiveEffect.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+import mage.abilities.Ability;
+import mage.abilities.common.delayed.ReflexiveTriggeredAbility;
+import mage.game.Game;
+import mage.target.targetpointer.FixedTarget;
+
+/**
+ * @author Susucr
+ */
+public class RegenerateSourceWithReflectiveEffect extends RegenerateSourceEffect {
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/effects/common/RegenerateSourceWithReflectiveEffect.java:11]: "reflective" -> "reflexive"

File: Mage/src/main/java/mage/abilities/effects/common/RegenerateSourceWithReflectiveEffect.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    private final ReflexiveTriggeredAbility reflexive;
+    private final boolean setReflectiveTarget;
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/effects/common/RegenerateSourceWithReflectiveEffect.java:14]: same here

**Review by xenohedron — APPROVED** [review]

**Pull request merged**